### ISSUE TITLE
Discrete logarithms

### ISSUE THREAD

**Issue by gorisaka** [sympy/sympy#11502]
Unless I'm missing something, it seems SymPy doesn't offer any way to compute discrete logarithms (i.e. generalized multiplicative order), only the multiplicative order.

For instance, Mathematica does it via ``MultiplicativeOrder[k,n,{r1,r2,…}]``.
http://reference.wolfram.com/language/ref/MultiplicativeOrder.html

What I would like to achieve in SymPy is something like:
```python
>>> from sympy.ntheory import discrete_log
>>> discrete_log(65537, 3**10, 3)
10
```
and ultimately
```python
>>> from sympy.polys.domains import GF
>>> GF(65537).log(3**10, 3)
10
>>> a = GF(65537)(3**11)
>>> a.log(3)
11
```

If not mistaken, I think this is what Sage implements natively in addition to use bindings to PARI/GP:
* Trial multiplication
* Shank's baby-step giant-step
* Pohlig-Hellman algorithm
* Pollard's rho algorithm (for logarithm)

These underlying algorithms would be chosen depending on the input (e.g. subgroup order size, prime order or not, etc). 

Before pushing any code, I would like your feedback.

Should SymPy implement this too?

Thanks.

**Comment by asmeurer** [comment]
Yes, this sounds like it would be a good addition.

**Comment by prateekg** [comment]
Yes, it will be good thing to start and I would be happy to contribute into it.

**Comment by gorisaka** [comment]
@prateekg Sure, you are welcomed to contribute.
